File: sprint8/By languages/go/lesson6code2.go
func prefixFunction(s string) []int {
    // Функция возвращает массив длины |s|
    π := make([]int, len(s))
    for i := 1; i < len(s); i++ {
        k := π[i-1]
        for k > 0 && s[k] != s[i] {
            k = π[k-1]
        }
        if s[k] == s[i] {
            k++
        }
        π[i] = k
    }
    return π
}

File: sprint8/By languages/swift/lesson6code2.swift
func prefixFunction(s: String) -> [Int] {
    // Функция возвращает массив длины |s|
    var π = [Int](repeating: 0, count: s.count)
    let sChars = Array(s)
    
    for i in 1..<s.count {
        var k = π[i-1]
        while k > 0 && sChars[k] != sChars[i] {
            k = π[k-1]
        }
        if sChars[k] == sChars[i] {
            k += 1
        }
        π[i] = k
    }
    return π
}
